core/lib/ReactCore_Internal.re

```diff
@@ -921,6 +933,303 @@ module Make = (Implementation: HostImplementation) => {
+          switch (newInstance.component.elementType) {
+          | React => assert(false)
+          | Host =>
+            newInstance.subElements.updateInstance(
+              Render.createSelf(~instance=newInstance),
+              parentHostView,
+            )
+          };
+          applyUpdateLogEntry(
+            ~subtreeChange,
+            ~parentHostView,
+            ~oldInstance,
+            ~newInstance,
+            (subtreeReactChange :> UpdateLog.subtreeChange),
+          );
+        };
+
+    let mountRenderedElement = (parentHostView, renderedElement) : unit => {
+      Implementation.beginChanges();
+      traverseRenderedElement(
+        parentHostView,
+        renderedElement,
+        mountInstanceUtil(Implementation.mountChild),
+        0,
+      );
+      Implementation.commitChanges();
+    };
+
+    let rec applyUpdateLogUtil =
+            (
+              ~subtreeChange: option(UpdateLog.subtreeChange),
+              parentHostView,
+              updateLog: list(UpdateLog.entry),
+            )
+            : unit =>
+      UpdateLog.(
+        switch (updateLog) {
+        | [
+            UpdateInstance({subTreeChanged, oldInstance, newInstance, _}),
+            ...tl,
+          ] =>
+          applyUpdateLogEntry(
+            ~subtreeChange,
+            ~parentHostView,
+            ~oldInstance,
+            ~newInstance,
+            subTreeChanged,
+          );
+          let parentHostView =
+            switch (subTreeChanged) {
+            | `Nested => getHostViewInstance(oldInstance)
+            | _ => parentHostView
+            };
+          applyUpdateLogUtil(~subtreeChange, parentHostView, tl);
+        | [
+            ChangeComponent({
+              oldOpaqueInstance: Instance(oldInstance),
+              newOpaqueInstance: Instance(newInstance),
+              _,
+            }),
+            ...tl,
+          ] =>
+          unmountInstance(~parentHostView, ~instance=oldInstance);
+          mountInstance(~parentHostView, ~instance=newInstance, ~position=0);
+          applyUpdateLogUtil(~subtreeChange, parentHostView, tl);
+        | [] => ()
+        }
+      );
+
+    let applyUpdateLog =
+        (
+          ~subtreeChange: option(UpdateLog.subtreeChange)=None,
+          parentHostView,
+          updateLog: list(UpdateLog.entry),
+        )
+        : unit => {
+      Implementation.beginChanges();
+      applyUpdateLogUtil(~subtreeChange, parentHostView, updateLog);
+      Implementation.commitChanges();
+    };
+
+    let applyTopLevelUpdate =
+        (
+          parentHostView: hostView,
+          renderedElement: renderedElement,
+          topLevelUpdate: option(RenderedElement.topLevelUpdate),
+        )
+        : unit =>
+      switch (topLevelUpdate) {
+      | Some({subtreeChange, updateLog}) when List.length(updateLog^) > 0 =>
+        let updateLog = List.rev(updateLog^);
+        applyUpdateLog(
```

Comment:
why can't we call `applyUpdateLogEntry with `subtreeChange` here, and then call `applyUpdateLog` with `updateLog`?